File: packages/crypto/src/vault-crypto.ts
import {
  encryptAES256GCM,
  decryptAES256GCM,
  generateRandomBytes,
  type EncryptedPayload,
} from './aes-gcm';

export type { EncryptedPayload };

export interface VaultKeyBundle {
  encryptedVaultKey: Uint8Array;
  nonce: Uint8Array;
  tag: Uint8Array;
}

const ENCODER = new TextEncoder();
const DECODER = new TextDecoder();
const VAULT_KEY_LENGTH = 32;

/**
 * Generate a random 32-byte vault key.
 * This key encrypts all vault items and is itself wrapped by the user's encKey.
 */
export function generateVaultKey(): Uint8Array {
  return generateRandomBytes(VAULT_KEY_LENGTH);
}

/**
 * Encrypt the vault key using the user's encryption key (derived from password).
 * Stored on the server; only decryptable with the user's encKey.
 */
export async function encryptVaultKey(
  encKey: Uint8Array,
  vaultKey: Uint8Array,
): Promise<VaultKeyBundle> {
  const payload = await encryptAES256GCM(encKey, vaultKey);
  return {
    encryptedVaultKey: payload.ciphertext,
    nonce: payload.nonce,
    tag: payload.tag,
  };
}

/**
 * Decrypt the vault key using the user's encryption key.
 */
export async function decryptVaultKey(
  encKey: Uint8Array,
  bundle: VaultKeyBundle,
): Promise<Uint8Array> {
  return decryptAES256GCM(encKey, {
    ciphertext: bundle.encryptedVaultKey,
    nonce: bundle.nonce,
    tag: bundle.tag,
  });
}

/**
 * Encrypt a vault item (JSON string) with the vault key.
 */
export async function encryptVaultItem(
  vaultKey: Uint8Array,
  plaintext: string,
): Promise<EncryptedPayload> {
  return encryptAES256GCM(vaultKey, ENCODER.encode(plaintext));
}

/**
 * Decrypt a vault item back to its JSON string.
 */
export async function decryptVaultItem(
  vaultKey: Uint8Array,
  payload: EncryptedPayload,
): Promise<string> {
  const bytes = await decryptAES256GCM(vaultKey, payload);
  return DECODER.decode(bytes);
}

File: apps/ios/AuthBoxCrypto/Sources/AuthBoxCrypto/VaultCrypto.swift
import Foundation

/// Vault encryption operations.
///
/// Mirrors `@authbox/crypto` vault-crypto.ts:
/// - Random 32-byte vault key
/// - Vault key wrapped with user's encKey (AES-256-GCM)
/// - Per-item encryption with vault key
public enum VaultCrypto {

    private static let vaultKeyLength = 32

    /// Generate a random 32-byte vault key.
    public static func generateVaultKey() -> Data {
        AES256GCM.generateRandomBytes(vaultKeyLength)
    }

    /// Encrypt the vault key using the user's encryption key.
    /// Stored on the server; only decryptable with the user's encKey.
    public static func encryptVaultKey(encKey: Data, vaultKey: Data) throws -> VaultKeyBundle {
        let payload = try AES256GCM.encrypt(key: encKey, plaintext: vaultKey)
        return VaultKeyBundle(
            encryptedVaultKey: payload.ciphertext,
            nonce: payload.nonce,
            tag: payload.tag
        )
    }

    /// Decrypt the vault key using the user's encryption key.
    public static func decryptVaultKey(encKey: Data, bundle: VaultKeyBundle) throws -> Data {
        try AES256GCM.decrypt(
            key: encKey,
            payload: EncryptedPayload(
                ciphertext: bundle.encryptedVaultKey,
                nonce: bundle.nonce,
                tag: bundle.tag
            )
        )
    }

    /// Encrypt a vault item (JSON string) with the vault key.
    public static func encryptVaultItem(vaultKey: Data, plaintext: String) throws -> EncryptedPayload {
        try AES256GCM.encrypt(key: vaultKey, plaintext: Data(plaintext.utf8))
    }

    /// Decrypt a vault item back to its JSON string.
    public static func decryptVaultItem(vaultKey: Data, payload: EncryptedPayload) throws -> String {
        let data = try AES256GCM.decrypt(key: vaultKey, payload: payload)
        guard let text = String(data: data, encoding: .utf8) else {
            throw VaultCryptoError.invalidUTF8
        }
        return text
    }
}

public enum VaultCryptoError: Error, Sendable {
    case invalidUTF8
}
